From ea3b8b18701659849e696c11cb06caed484eadb1 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Sat, 23 Mar 2024 20:37:39 +0100
Subject: [PATCH 1033/1141] media: dcmipp: isp: fix set_selection handling

Depending on the which parameter, either the dcmipp structure
src_fmt or the try_format should be updated.
Moreover the decimate variable should only be updated when
performing ACTIVE setting.

Change-Id: I5eb9bcf354a7f0da9ca07a1fe8efd4c33f1f944e
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/369266
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-isp.c        | 33 +++++--------------
 1 file changed, 9 insertions(+), 24 deletions(-)

--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
@@ -465,7 +465,7 @@ static int dcmipp_isp_set_selection(stru
 				    struct v4l2_subdev_selection *s)
 {
 	struct dcmipp_isp_device *isp = v4l2_get_subdevdata(sd);
-	struct v4l2_mbus_framefmt *sink_fmt;
+	struct v4l2_mbus_framefmt *sink_fmt, *src_fmt;
 	struct v4l2_rect *crop, *compose;
 	unsigned int dec;
 
@@ -474,10 +474,12 @@ static int dcmipp_isp_set_selection(stru
 
 	if (s->which == V4L2_SUBDEV_FORMAT_ACTIVE) {
 		sink_fmt = &isp->sink_fmt;
+		src_fmt = &isp->src_fmt;
 		crop = &isp->crop;
 		compose = &isp->compose;
 	} else {
 		sink_fmt = v4l2_subdev_get_try_format(sd, state, s->pad);
+		src_fmt = v4l2_subdev_get_try_format(sd, state, 1);
 		crop = v4l2_subdev_get_try_crop(sd, state, s->pad);
 		compose = v4l2_subdev_get_try_compose(sd, state, s->pad);
 	}
@@ -490,20 +492,12 @@ static int dcmipp_isp_set_selection(stru
 
 		/* When we set the crop, this impact as well the compose */
 		*compose = s->r;
-		isp->decimation = 0;
 
-		/*
-		 * In case of setting the crop with ACTIVE set, we need to
-		 * update the source pad size
-		 */
-		if (s->which == V4L2_SUBDEV_FORMAT_ACTIVE) {
-			isp->src_fmt.width = s->r.width;
-			isp->src_fmt.height = s->r.height;
-		}
+		src_fmt->width = s->r.width;
+		src_fmt->height = s->r.height;
 
-		/* TODO - when not in format active, we should also update the
-		 * try src pad format
-		 */
+		if (s->which == V4L2_SUBDEV_FORMAT_ACTIVE)
+			isp->decimation = 0;
 
 		dev_dbg(isp->dev, "s_selection: crop %ux%u@(%u,%u)\n",
 			crop->width, crop->height, crop->left, crop->top);
@@ -533,17 +527,8 @@ static int dcmipp_isp_set_selection(stru
 
 		*compose = s->r;
 
-		/*
-		 * In case of setting the compose with ACTIVE set, we need to
-		 * update the source pad size
-		 */
-		if (s->which == V4L2_SUBDEV_FORMAT_ACTIVE) {
-			isp->src_fmt.width = s->r.width;
-			isp->src_fmt.height = s->r.height;
-		}
-		/* TODO - when not in format active, we should also update the
-		 * try src pad format
-		 */
+		src_fmt->width = s->r.width;
+		src_fmt->height = s->r.height;
 
 		dev_dbg(isp->dev, "s_selection: compose %ux%u@(%u,%u)\n",
 			s->r.width, s->r.height, s->r.left, s->r.top);
